// open-holiday-js/dist/index.mjs
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/schemas/country.ts
import { z as z2 } from "zod";

// src/schemas/utils.ts
import { z } from "zod";
var isCode = (code) => /^[A-Z]{2}$/.test(code);
var isDate = (string) => /^\d{4}-\d{2}-\d{2}$/.test(string);
var nameSchema = z.object({
  language: z.string().refine(isCode),
  text: z.string()
});

// src/schemas/country.ts
var countrySchema = z2.object({
  isoCode: z2.string().refine(isCode),
  name: z2.array(nameSchema),
  officialLanguages: z2.array(z2.string().refine(isCode))
});
var countryApiResponse = z2.array(countrySchema);

// src/schemas/holidays.ts
import { z as z3 } from "zod";
var holidayType = z3.union([
  z3.literal("Public"),
  z3.literal("Bank"),
  z3.literal("National"),
  z3.literal("Regional"),
  z3.literal("Local"),
  z3.literal("School"),
  z3.literal("BackToSchool"),
  z3.literal("EndOfLessons")
]);
var holidaySchema = z3.object({
  id: z3.string().min(1),
  startDate: z3.string().refine(isDate),
  endDate: z3.string().refine(isDate),
  type: holidayType,
  name: z3.array(nameSchema),
  nationwide: z3.boolean(),
  subdivisions: z3.optional(
    z3.array(
      z3.object({
        code: z3.string().min(1),
        shortName: z3.string().min(1)
      })
    )
  ),
  quality: z3.optional(z3.union([z3.literal("Mandatory"), z3.literal("Optional")])),
  comment: z3.optional(z3.array(nameSchema))
});
var holidayApiResponse = z3.array(holidaySchema);

// src/schemas/languages.ts
import { z as z4 } from "zod";
var languageSchema = z4.object({
  isoCode: z4.string().refine(isCode),
  name: z4.array(nameSchema)
});
var languagesApiResponse = z4.array(languageSchema);

// src/schemas/subdivisions.ts
import { z as z5 } from "zod";
var subdivisionSchema = z5.object({
  code: z5.string().min(1),
  isoCode: z5.string(),
  shortName: z5.string().min(1),
  category: z5.array(nameSchema),
  name: z5.array(nameSchema),
  officialLanguages: z5.array(z5.string().refine(isCode))
});
var subdivisionsApiResponse = z5.array(subdivisionSchema);

// src/utils/date.ts
var stringifyDate = (date) => {
  const year = date.getFullYear();
  const month = date.getMonth() + 1;
  const day = date.getDate();
  return `${year}-${month}-${day}`;
};
var parseDate = (date) => {
  const [year, month, day] = date.split("-").map((n) => parseInt(n));
  return new Date(year, month - 1, day);
};

// src/utils/request.ts
function requestJson(url, schema) {
  return __async(this, null, function* () {
    const raw = yield fetch(url, { headers: { "accept": "text/json" } });
    if (!raw.ok)
      throw new Error(`Request was not successful (${raw.status})`);
    const unparsedContent = yield raw.json();
    return schema.parse(unparsedContent);
  });
}

// src/utils/url.ts
var serializeUrl = (url) => {
  const endsWithSlash = url.endsWith("/");
  if (endsWithSlash)
    return url.slice(0, -1);
  return url;
};
var createLangQuery = (language, isFirstParam = true) => {
  const param = language ? `languageIsoCode=${language.toUpperCase()}` : "";
  if (isFirstParam)
    return "?" + param;
  return param;
};

// src/index.ts
var Holiday = class {
  /**
   * Constructs a new instance of the Holiday SDK.
   * @param {HolidayOptions} options - The options for configuring the SDK.
   * @param {string} options.url - The base URL of the Open Holiday API; defaults to `https://openholidaysapi.org`
   */
  constructor({ url } = {}) {
    this.url = serializeUrl(url || "https://openholidaysapi.org");
  }
  /**
   * Get all countries
   *
   * [Official docs](https://www.openholidaysapi.org/en/#countries)
   *
   * @param language The optional language to get the country names in.
   *                 If a incorrect is given the data will be in english.
   *                 If no language is given the data will be in all available languages.
   *
   * @returns A promise that resolves in all countries supported.
   */
  getCountries(language) {
    const param = createLangQuery(language);
    const url = `${this.url}/Countries${param}`;
    return requestJson(url, countryApiResponse);
  }
  /**
   * Get all languages
   *
   * @param language The optional language to get the languages names in.
   *                 If a incorrect is given the data will be in english.
   *                 If no language is given the data will be in all available languages.
   *
   * @returns A promise that resolves in all langauges supported.
   */
  getLangs(language) {
    const param = createLangQuery(language);
    const url = `${this.url}/Languages${param}`;
    return requestJson(url, languagesApiResponse);
  }
  /**
   * Get all subdivisions in a state
   *
   * [Official docs](https://www.openholidaysapi.org/en/#subdivisions)
   *
   * @param country The iso-code of the country which subdivisions are in question
   * @param language The language in which to receive the subdivison names
   * @returns A promise that resolves into a list of all subdivions that are important for holidays
   */
  getSubdivisions(country, language) {
    const langParam = createLangQuery(language, false);
    const countryParam = `countryIsoCode=${country}`;
    const params = `?${langParam}&${countryParam}`;
    const url = `${this.url}/Subdivisions${params}`;
    return requestJson(url, subdivisionsApiResponse);
  }
  /**
   * Get school holidays
   *
   * [Official docs](https://www.openholidaysapi.org/en/public-holidays)
   *
   * @param country - The iso code of the country
   * @param startDate - The date to beginn the search from
   * @param endDate - The date to end the search at, maximum three years after startDate
   * @param subdivision - Optional administrative unit
   * @param language - The optional language to get all the data in
   *
   * @returns A promise that resolves into the public holidays
   */
  getPublicHolidays(country, startDate, endDate, subdivision, language) {
    return __async(this, null, function* () {
      const paramsArray = [
        createLangQuery(language, false),
        subdivision ? `subdivisionCode=${subdivision}` : "",
        `validFrom=${stringifyDate(startDate)}`,
        `validTo=${stringifyDate(endDate)}`,
        `countryIsoCode=${country}`
      ].filter((s) => s);
      const params = `?${paramsArray.join("&")}`;
      const url = `${this.url}/PublicHolidays${params}`;
      const data = yield requestJson(url, holidayApiResponse);
      const jsNative = data.map((d) => __spreadProps(__spreadValues({}, d), {
        startDate: parseDate(d.startDate),
        endDate: parseDate(d.endDate)
      }));
      return jsNative;
    });
  }
  /**
   * Get school holidays
   *
   * [Official docs](https://www.openholidaysapi.org/en/#school-holidays)
   *
   * @param country - The iso code of the country
   * @param startDate - The date to beginn the search from
   * @param endDate - The date to end the search at, maximum three years after startDate
   * @param subdivision - Optional administrative unit
   * @param language - The optional language to get all the data in
   *
   * @returns A promise that resolves into the school holidays
   */
  getSchoolHolidays(country, startDate, endDate, subdivision, language) {
    return __async(this, null, function* () {
      const paramsArray = [
        createLangQuery(language, false),
        subdivision ? `subdivisionCode=${subdivision}` : "",
        `validFrom=${stringifyDate(startDate)}`,
        `validTo=${stringifyDate(endDate)}`,
        `countryIsoCode=${country}`
      ].filter((s) => s);
      const params = `?${paramsArray.join("&")}`;
      const url = `${this.url}/SchoolHolidays${params}`;
      const data = yield requestJson(url, holidayApiResponse);
      const jsNative = data.map((d) => __spreadProps(__spreadValues({}, d), {
        startDate: parseDate(d.startDate),
        endDate: parseDate(d.endDate)
      }));
      return jsNative;
    });
  }
};
export {
  Holiday
};
